skimage/morphology/_max_tree.pyx

```diff
@@ -0,0 +1,459 @@
+
+    return 1
+
+
+# computes the area of all max-tree components
+# attribute to be used in area opening and closing
+cpdef np.ndarray[DTYPE_FLOAT64_t, ndim = 1] _compute_area(dtype_t[::1] image,
+                                                          DTYPE_INT64_t[::1] parent,
+                                                          DTYPE_INT64_t[::1] sorted_indices):
+    cdef DTYPE_INT64_t p_root = sorted_indices[0]
+    cdef DTYPE_INT64_t p, q
+    cdef DTYPE_UINT64_t number_of_pixels = len(image)
+    cdef np.ndarray[DTYPE_FLOAT64_t, ndim = 1] area = np.ones(number_of_pixels,
+                                                              dtype=np.float64)
+
+    for p in sorted_indices[::-1]:
+        if p == p_root:
+            continue
+        q = parent[p]
+        area[q] = area[q] + area[p]
+
+    return area
+
+
+# computes the bounding box extension of all max-tree components
+# attribute to be used in diameter opening and closing
+cpdef np.ndarray[DTYPE_FLOAT64_t, ndim = 1] _compute_extension(dtype_t[::1] image,
+                                                               DTYPE_INT32_t[::1] shape,
+                                                               DTYPE_INT64_t[::1] parent,
+                                                               DTYPE_INT64_t[::1] sorted_indices):
+    cdef DTYPE_INT64_t p_root = sorted_indices[0]
+    cdef DTYPE_INT64_t p, q
+    cdef DTYPE_UINT64_t number_of_pixels = len(image)
+    cdef np.ndarray[DTYPE_FLOAT64_t, ndim = 1] extension = np.ones(number_of_pixels,
+                                                                   dtype=np.float64)
+    cdef np.ndarray[DTYPE_FLOAT64_t, ndim = 2] max_coord = np.array(np.unravel_index(np.arange(number_of_pixels),
+                                                                                   shape), dtype=np.float64).T
+    cdef np.ndarray[DTYPE_FLOAT64_t, ndim = 2] min_coord = np.array(np.unravel_index(np.arange(number_of_pixels),
+                                                                                   shape), dtype=np.float64).T
+
+    for p in sorted_indices[::-1]:
+        if p == p_root:
+            continue
+        q = parent[p]
+        max_coord[q] = np.maximum(max_coord[q], max_coord[p])
+        min_coord[q] = np.minimum(min_coord[q], min_coord[p])
+        extension[q] = np.max(max_coord[q] - min_coord[q]) + 1
+
+    return extension
+
+
+# direct filter (criteria based filter)
+cpdef void _direct_filter(dtype_t[::1] image,
+                          dtype_t[::1] output,
+                          DTYPE_INT64_t[::1] parent,
+                          DTYPE_INT64_t[::1] sorted_indices,
+                          DTYPE_FLOAT64_t[::1] attribute,
+                          DTYPE_FLOAT64_t attribute_threshold
+                          ):
+    """Direct filtering. Produces an image in which for all possible
+    thresholds, each connected component has an
+    attribute >= attribute_threshold.
```

Comment:
Can you explain this, maybe with an example? I am having a hard time visualising what is happening...

Reply:
This is actually not that easy. It is the primitive which is called for instance from area_opening, diameter_opening. If you apply an area_opening with a certain area_threshold, let's say 5, you have the guarantee that in the resulting image all connected components for all possible thresholds have at least an area of 5. 

The problem is that it is not very intuitive if you formulate that for general attributes. The biggest problem are non-increasing attributes, like circularity. But at the moment, I did not include any example for this.

Reply:
> If you apply an area_opening with a certain area_threshold, let's say 5, you have the guarantee that in the resulting image all connected components for all possible thresholds have at least an area of 5.

You're blowing my mind. =D